src/formula/callable_objects.cpp

```diff
@@ -162,7 +162,20 @@ int attack_type_callable::do_compare(const formula_callable* callable) const
 		return att_->range().compare(att_callable->att_->range());
 	}
 
-	return att_->weapon_specials().compare(att_callable->att_->weapon_specials());
+	const auto self_specials = att_->specials().all_children_range();
+	const auto other_specials = att_callable->att_->specials().all_children_range();
+	if(self_specials.size() != other_specials.size()) {
+		return self_specials.size() < other_specials.size() ? -1 : 1;
+	}
+	for(std::size_t i = 0; i < self_specials.size(); ++i) {
+		const auto& s = self_specials[i].cfg["id"];
+		const auto& o = other_specials[i].cfg["id"];
+		if(s != o) {
+			return s.str() < o.str() ? -1 : 1;
```

Comment:
Any reason not to just `return s.compare(o)`?

Reply:
Will change to `s.str().compare(o.str())`, thanks. They're both config_attribute_value objects, so there isn't a direct compare(), but I forgot to look again for a compare() function after realising that I needed to convert them to strings.

Reply:
I realize this is a bit late now that it's merged, but you could also do:
```cpp
auto result = s.str().compare(o.str());
if(result != 0) return result;
```

Reply:
Good idea, I'll remember that next time.